src/mame/drivers/didact.cpp

```diff
@@ -457,34 +451,381 @@ void mp68a_state::machine_start()
+	AM_RANGE(0xc800, 0xc81f) AM_READWRITE(pia_r, pia_w) AM_MIRROR(0x07e0)
+	AM_RANGE(0xd000, 0xffff) AM_ROM AM_REGION("roms", 0x1000)
 ADDRESS_MAP_END
 
 // This address map is traced from schema
 static ADDRESS_MAP_START( md6802_map, AS_PROGRAM, 8, md6802_state )
 	AM_RANGE(0x0000, 0x07ff) AM_RAM AM_MIRROR(0x1800)
-	AM_RANGE(0xa000, 0xa003) AM_DEVREADWRITE("pia1", pia6821_device, read, write) AM_MIRROR(0x1ffc)
-	AM_RANGE(0xc000, 0xc003) AM_DEVREADWRITE("pia2", pia6821_device, read, write) AM_MIRROR(0x1ffc)
+	AM_RANGE(0xa000, 0xa003) AM_DEVREADWRITE(PIA1_TAG, pia6821_device, read, write) AM_MIRROR(0x1ffc)
+	AM_RANGE(0xc000, 0xc003) AM_DEVREADWRITE(PIA2_TAG, pia6821_device, read, write) AM_MIRROR(0x1ffc)
 	AM_RANGE(0xe000, 0xe7ff) AM_ROM AM_MIRROR(0x1800) AM_REGION("maincpu", 0xe000)
 ADDRESS_MAP_END
 
 // This address map is traced from pcb
 static ADDRESS_MAP_START( mp68a_map, AS_PROGRAM, 8, mp68a_state )
 	AM_RANGE(0x0000, 0x00ff) AM_RAM AM_MIRROR(0xf000)
-	AM_RANGE(0x0500, 0x0503) AM_DEVREADWRITE("pia1", pia6820_device, read, write) AM_MIRROR(0xf0fc)
-	AM_RANGE(0x0600, 0x0603) AM_DEVREADWRITE("pia2", pia6820_device, read, write) AM_MIRROR(0xf0fc)
+	AM_RANGE(0x0500, 0x0503) AM_DEVREADWRITE(PIA1_TAG, pia6820_device, read, write) AM_MIRROR(0xf0fc)
+	AM_RANGE(0x0600, 0x0603) AM_DEVREADWRITE(PIA2_TAG, pia6820_device, read, write) AM_MIRROR(0xf0fc)
 	AM_RANGE(0x0700, 0x07ff) AM_RAM AM_MIRROR(0xf000)
 	AM_RANGE(0x0800, 0x0bff) AM_ROM AM_MIRROR(0xf400) AM_REGION("maincpu", 0x0800)
 ADDRESS_MAP_END
 
-/* Input ports */
+/* Input ports
+ * Four e100 keys are not mapped yet,
+ * - The redundant '*' on the keyboard together with the '\'' single quote, both on same e100 key
+ * - The 'E' key on the keypad, presumably used for calculator applications to remove the last entered number
+ * - The 'Break' key will be mapped to NMI at some point
+ * - The 'REPT' key has a so far unknown function
+ */
 static INPUT_PORTS_START( e100 )
+/*	Bits read on PIA1 A when issueing line number on PIA1 B bits 0-3 through a 74145 demultiplexer */
+	PORT_START("LINE0")
+	PORT_BIT(0x01, IP_ACTIVE_LOW,	IPT_UNUSED)	  /* Port name      key pressed on US/UK keyboard   */
+	PORT_BIT(0x02, IP_ACTIVE_LOW,	IPT_UNUSED)
+	PORT_BIT(0x04, IP_ACTIVE_LOW,	IPT_KEYBOARD) PORT_NAME("SPACE")  PORT_CODE(KEYCODE_SPACE)		PORT_CHAR(' ')
+	PORT_BIT(0x08, IP_ACTIVE_LOW,	IPT_KEYBOARD) PORT_NAME("LSHIFT") PORT_CODE(KEYCODE_LSHIFT)		PORT_CODE(KEYCODE_RSHIFT)		PORT_CHAR(UCHAR_SHIFT_1)
+	PORT_BIT(0x10, IP_ACTIVE_LOW,	IPT_KEYBOARD) PORT_NAME("LCTRL")  PORT_CODE(KEYCODE_LCONTROL)	PORT_CHAR(UCHAR_SHIFT_2)
+	PORT_BIT(0x20, IP_ACTIVE_LOW,	IPT_KEYBOARD) PORT_NAME("REPT") /* Not mapped yet */
+	PORT_BIT(0x40, IP_ACTIVE_LOW,	IPT_UNUSED)
+	PORT_BIT(0x80, IP_ACTIVE_LOW,	IPT_UNUSED)
+
+	PORT_START("LINE1")
+	PORT_BIT(0x01, IP_ACTIVE_LOW,	IPT_UNUSED)
+	PORT_BIT(0x02, IP_ACTIVE_LOW,	IPT_UNUSED)
+	PORT_BIT(0x04, IP_ACTIVE_LOW,	IPT_KEYBOARD) PORT_NAME("Z")  PORT_CODE(KEYCODE_Z)  PORT_CHAR('Z')
```

Comment:
You don't need PORT_NAME for keyboard keys where the name is the same as the keycap label.  You should also supply shifted and unshifted variants (assuming the machine supports lowercase).  This key should probably be:

    PORT_BIT(0x04, IP_ACTIVE_LOW,	IPT_KEYBOARD) PORT_CODE(KEYCODE_Z) PORT_CHAR('z') PORT_CHAR('Z')

And similar for other keys